doc/src/devdocs/llvm-passes.md

```diff
@@ -0,0 +1,161 @@
+* Opt Name `function(DemoteFloat16)`
+
+This pass replaces [float16](https://en.wikipedia.org/wiki/Half-precision_floating-point_format) operations with float32 operations on architectures that do not natively support float16 operations. This is done by inserting `fpext` and `fptrunc` instructions around any float16 operation. On architectures that do support native float16 operations, this pass is a no-op.
+
+### LateGCLowering
+
+* Filename: `llvm-late-gc-lowering.cpp`
+* Class Name: `LateLowerGCPass`
+* Opt Name: `function(LateLowerGCFrame)`
+
+This pass performs most of the GC rooting work required to track pointers between GC safepoints. It also lowers several intrinsics to their corresponding instruction translation, and is permitted to violate the non-integral invariants previously established (`pointer_from_objref` is lowered to a `ptrtoint` instruction here). This pass typically occupies the most time out of all the custom Julia passes, due to its dataflow algorithm to minimize the number of objects live at any safepoint.
+
+### FinalGCLowering
+
+* Filename: `llvm-final-gc-lowering.cpp`
+* Class Name: `FinalLowerGCPass`
+* Opt Name: `module(FinalLowerGC)`
+
+This pass lowers a few last intrinsics to their final form targeting functions in the `libjulia` library. Separating this from `LateGCLowering` enables other backends (GPU compilation) to supply their own custom lowerings for these intrinsics, enabling the Julia pipeline to be used on those backends as well.
+
+### LowerHandlers
+
+* Filename: `llvm-lower-handlers.cpp`
+* Class Name: `LowerExcHandlersPass`
+* Opt Name: `function(LowerExcHandlers)`
+
+This pass lowers exception handling intrinsics into calls to runtime functions that are actually called when handling exceptions.
+
+### RemoveNI
+
+* Filename: `llvm-remove-ni.cpp`
+* Class Name: `RemoveNIPass`
+* Opt Name: `module(RemoveNI)`
+
+This pass removes the non-integral address spaces from the module's datalayout string. This enables the backend to lower Julia's custom address spaces directly to machine code, without a costly rewrite of every pointer operation to address space 0.
+
+### SIMDLoop
+
+* Filename: `llvm-simdloop.cpp`
+* Class Name: `LowerSIMDLoopPass`
+* Opt Name: `module(LowerSIMDLoop)`
+
+This pass acts as the main driver of the `@simd` annotation. Codegen inserts a call to a marker intrinsic (`julia.simdloop`), which this pass uses to identify loops that were originally marked with `@simd`. Then, this pass looks for a chain of floating point operations that form a reduce and marks adds fast math to allow reassociation (and thus vectorization). This pass does not preserve either loop information nor inference correctness, so it may violate Julia semantics in surprising ways. If the loop was annotated with `ivdep` as well, then the pass marks the loop as having no loop-carried dependencies (the resulting behavior is undefined if the user annotation was incorrect or gets applied to the wrong loop).
```

Comment:
```suggestion
This pass acts as the main driver of the `@simd` annotation. Codegen inserts a call to a marker intrinsic (`julia.simdloop`), which this pass uses to identify loops that were originally marked with `@simd`. Then, this pass looks for a chain of floating point operations that form a reduce and adds the `contract` and `reassoc` fast math flags to allow reassociation (and thus vectorization). This pass does not preserve either loop information nor inference correctness, so it may violate Julia semantics in surprising ways. If the loop was annotated with `ivdep` as well, then the pass marks the loop as having no loop-carried dependencies (the resulting behavior is undefined if the user annotation was incorrect or gets applied to the wrong loop).
```